### PR TITLE
[88] Disband groups with unavailable suite sizes on starting lottery

### LINKED ISSUE #88: Handle group sizes when a suite size becomes unavailable
Group sizes are restricted to suite sizes available in the draw at the time of group creation/editing, but those sizes may become unavailable after the group sets its size.

We need to figure out how to handle this situation. I think that an email notification to the affected groups that tells them they have to change their size along with a banner notification for the group leader while signed in would work well.

### PR DESCRIPTION
**Pull request by esoterik**
Resolves #88

### CODE REVIEW

File: app/models/draw.rb [hunk 1]
```diff
@@ -34,6 +34,13 @@ def suite_sizes
     SuiteSizesQuery.new(suites.available).call
   end
 
+  # Finds the sizes that groups exist for within the draw
+  #
+  # @return [Array<Integer>] the available group sizes
+  def group_sizes
```
Comment (orenyk) [inline app/models/draw.rb]: We should memoize this and use it in `oversubscribed?`.

**Comment by orenyk** [comment]
Bigger issue - if you want to actually set up this scenario then draw#show doesn't render. Steps to replicate:

1. Add group of a given size to a draw
2. Remove all suites of that size from draw
3. Return to draw#show

It's probably because of #278, so maybe we just merge that into here?

**Comment by esoterik** [comment]
@orenyk  ready for review! now resolves #278 as well as #88

File: app/controllers/draws_controller.rb [hunk 2]
```diff
@@ -183,15 +184,17 @@ def calculate_group_metrics
...
       empty_groups_hash.merge(@groups.sort_by { |g| Group.statuses[g.status] }
                                      .group_by(&:size))
+    @groups_by_size.default = 0
+    @group_sizes = @groups_by_size.keys
   end
 
   def calculate_oversub_metrics # rubocop:disable AbcSize
     return unless policy(@draw).oversub_report?
-    zeroed_group_hash = @suite_sizes.map { |s| [s, 0] }.to_h
+    zeroed_group_hash = @group_sizes.map { |s| [s, 0] }.to_h
```
Comment (orenyk) [inline app/controllers/draws_controller.rb]: I feel like we want to use `@group_sizes + @suite_sizes` here, otherwise we won't show suite sizes with no groups. I think that was originally my point in #278, sorry if it wasn't clear.

File: spec/models/draw_spec.rb [hunk 3]
```diff
@@ -44,6 +44,17 @@
...
   end
 
+  describe '#group_sizes' do
+    let(:draw) { FactoryGirl.build_stubbed(:draw) }
+    it 'returns an array of group sizes in the draw' do
+      instance_spy('GroupSizesQuery', call: [1, 2]).tap do |q|
+        allow(GroupSizesQuery).to receive(:new).with(draw.suites.available)
+          .and_return(q)
+      end
+      expect(draw.group_sizes).to match_array([1, 2])
```
Comment (orenyk) [inline spec/models/draw_spec.rb]: We can just assert that `GroupSizesQuery` gets called, we don't need to test against return values.

File: spec/models/draw_spec.rb [hunk 4]
```diff
@@ -44,6 +44,17 @@
     end
   end
 
+  describe '#group_sizes' do
+    let(:draw) { FactoryGirl.build_stubbed(:draw) }
+    it 'returns an array of group sizes in the draw' do
+      instance_spy('GroupSizesQuery', call: [1, 2]).tap do |q|
+        allow(GroupSizesQuery).to receive(:new).with(draw.suites.available)
```
Comment (orenyk) [inline spec/models/draw_spec.rb]: Snake case, just to be safe.

File: app/controllers/draws_controller.rb (hunk 2 above)
Comment (orenyk) [inline app/controllers/draws_controller.rb]: Also, just to be clear, this makes `calculate_oversub_metrics` dependent on `calculate_group_metrics`. It's fine, but we might want to comment that somewhere?

File: app/helpers/draws_helper.rb [hunk 5]
```diff
@@ -14,7 +14,7 @@ def intent_deadline_str(draw)
   end
 
   def diff_class(diff)
-    raise ArgumentError unless diff.is_a? Integer
+    return 'negative' unless diff.is_a? Integer
```
Comment (orenyk) [inline app/helpers/draws_helper.rb]: ... I don't like this. I _am_ ok with returning `''` if we don't pass an integer.

File: app/controllers/draws_controller.rb [hunk 6]
```diff
@@ -174,6 +174,7 @@ def calculate_metrics
   def calculate_suite_metrics
     @suite_sizes = SuiteSizesQuery.new(@draw.suites.available).call
     @suite_counts = @draw.suites.available.group(:size).count
+    @suite_counts.default = 0
```
Comment (orenyk) [inline app/controllers/draws_controller.rb]: Oo awesome - does this mean we can rid of `zeroed_group_hash` and `empty_groups_hash`? Didn't realize this was a thing, very cool.

File: app/helpers/draws_helper.rb (hunk 5 above)
Comment (esoterik) [inline app/helpers/draws_helper.rb]: oh, yeah this isn't even necessary any more -- I'll fix it

File: app/controllers/draws_controller.rb (hunk 2 above)
Comment (esoterik) [inline app/controllers/draws_controller.rb]: the oversub report still shows suites with no groups as-is

Comment (esoterik) [inline app/controllers/draws_controller.rb]: it also shows sizes for which there are no suites but there are groups

Comment (orenyk) [inline app/controllers/draws_controller.rb]: Oh ok, I'm not sure how I didn't see that but :+1:

Comment (orenyk) [inline app/controllers/draws_controller.rb]: Ohhh, now I see. Ok, I think that `@group_sizes` is a confusing name then - it implies that it's only the group sizes. Can we make it clearer?

File: app/controllers/draws_controller.rb [hunk 7]
```diff
@@ -165,36 +165,39 @@ def set_draw
...
-    @group_counts = zeroed_group_hash.merge(@groups.group(:size).count)
-    @locked_counts = zeroed_group_hash.merge(@groups.where(status: 'locked')
-                                                    .group(:size).count)
-    @diff = @group_sizes.map do |size|
+    calculate_sizes
+    @group_counts = @groups.group(:size).count
+    @group_counts.default = 0
+    @locked_counts = @groups.where(status: 'locked').group(:size).count
+    @locked_counts.default = 0
+    @diff = @sizes.map do |size|
```
Comment (orenyk) [inline app/controllers/draws_controller.rb]: Minor,  but do we want to replace this with `each_with_object` to avoid `to_h`? Now that we've found it I think it reads much better.

File: spec/features/draws/draw_oversubscription_handling_spec.rb [hunk 8]
```diff
@@ -23,6 +23,8 @@
     expect(page).to have_content('Oversubscription handling')
   end
 
+  # TODO: figure out why this spec fails to send sizes to the
```
Comment (orenyk) [inline spec/features/draws/draw_oversubscription_handling_spec.rb]: Is this still broken?

File: spec/models/draw_spec.rb [hunk 9]
```diff
@@ -47,11 +47,12 @@
   describe '#group_sizes' do
     let(:draw) { FactoryGirl.build_stubbed(:draw) }
     it 'returns an array of group sizes in the draw' do
-      instance_spy('GroupSizesQuery', call: [1, 2]).tap do |q|
+      g = instance_spy('group_sizes_query', call: [1, 2]).tap do |q|
```
Comment (orenyk) [inline spec/models/draw_spec.rb]: We don't need to specify a return for `call` here.

File: app/views/draws/_group_report.html.erb [hunk 10]
```diff
@@ -1,7 +1,8 @@
 <% if groups.all? { |_s, groups| groups.empty? } %>
   <p>There are no groups yet.</p>
 <% else %>
-  <% sizes.each do |size| %>
+  <% # THIS IS THE HACKY FIX %>
```
Comment (orenyk) [inline app/views/draws/_group_report.html.erb]: ?? Was this fixing the issue in the spec before?

File: app/views/draws/_oversub_report.html.erb [hunk 11]
```diff
@@ -11,7 +11,7 @@
     </tr>
   </thead>
   <tbody>
-    <% @suite_sizes.each do |size| %>
+    <% @sizes.each do |size| %>
```
Comment (orenyk) [inline app/views/draws/_oversub_report.html.erb]: Are we not currently sending in local variables to this partial? Fine to leave if not.

File: spec/models/draw_spec.rb [hunk 12]
```diff
@@ -44,6 +44,18 @@
     end
   end
 
+  describe '#group_sizes' do
+    let(:draw) { FactoryGirl.build_stubbed(:draw) }
+    it 'returns an array of group sizes in the draw' do
+      g = instance_spy('group_sizes_query', call: [1, 2]).tap do |q|
```
Comment (orenyk) [inline spec/models/draw_spec.rb]: I don't think we need to specify the return of `call` here anymore.

**Comment by orenyk** [comment]
Also, rubocop 😛

File: app/views/draws/_group_report.html.erb (hunk 10 above)
Comment (orenyk) [inline app/views/draws/_group_report.html.erb]: Found it - you need to update `app/views/draws/oversubscription.html.erb` as well to pass `@sizes` instead of `@draw_sizes`.

File: spec/features/draws/draw_oversubscription_handling_spec.rb (hunk 8 above)
Comment (orenyk) [inline spec/features/draws/draw_oversubscription_handling_spec.rb]: See [my other comment](https://github.com/YaleSTC/vesta/pull/384#discussion_r106812467).

File: spec/models/draw_spec.rb (hunk 9 above)
Comment (esoterik) [inline spec/models/draw_spec.rb]: I'm not sure that we can just test that the method was called on the query object here -- we be ensuring that the return value of `group_sizes` is correct, so we need to set a return value for `call`

File: app/controllers/draws_controller.rb (hunk 7 above)
Comment (esoterik) [inline app/controllers/draws_controller.rb]: I think it's ok for now, plus we will be refactoring this later

File: spec/features/draws/draw_oversubscription_handling_spec.rb (hunk 8 above)
Comment (orenyk) [inline spec/features/draws/draw_oversubscription_handling_spec.rb]: Can we remove this comment?

File: spec/models/draw_spec.rb [hunk 13]
```diff
@@ -161,18 +172,6 @@
     end
   end
 
-  describe '#all_intents_declared?' do
```
Comment (orenyk) [inline spec/models/draw_spec.rb]: Why did you remove this test?

Comment (esoterik) [inline spec/models/draw_spec.rb]: it was a duplicate

**Comment by orenyk** [comment]
PR failure is intermittent and fixed in #410, merging.
